Build warning "warning: pointer ‘where’ may be used after ‘realloc’" in util.c

Recent builds are provoking a gcc warning:

```
In file included from util.c:26:
util.c: In function ‘Perl_safesysrealloc’:
util.c:329:17: warning: pointer ‘where’ may be used after ‘realloc’ [-Wuse-after-free]
  329 |         DEBUG_m(PerlIO_printf(Perl_debug_log, "0x%" UVxf ": (%05ld) rfree\n",PTR2UV(where),(long)PL_an++));
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
perl.h:4899:37: note: in definition of macro ‘DEBUG_m’
 4899 |                                     a;                                  \
      |                                     ^
util.c:290:15: note: call to ‘realloc’ here
  290 |         ptr = (Malloc_t)PerlMem_realloc(where,size);
```

Having looked into the code a bit I don't think this is really actually a problem. We're using a pointer after realloc, but only to convert it into a numerical representation to print it in a debugging message. This should be fine.

To fix it: Either we could convert the number into a UV before realloc, or we could somehow quiet the warning on this one line - though I don't know how to do that.

> PR #20607 should fix this

Unfortunately, it does not appear to do so.  Configuring as follows on FreeBSD-12:
```
$ git checkout leonerd-avoid-warning-use-after-free
Switched to branch 'leonerd-avoid-warning-use-after-free'
[perlmonger: perl] $ sh ./Configure -des -Dusedevel -Duseithreads -DDEBUGGING -Dcc=gcc13 && \
> make miniperl 2>&1 | tee $P5P_DIR/make-output/`extract-sha`.freebsd.threaded.gcc13.debugging.leonerd.makeminiperl.output.txt
```
... I observed the following in the output of `make miniperl`:
```
gcc13 -c -DPERL_CORE -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_FORTIFY_SOURCE=2 -std=c99 -O2 -g -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wno-declaration-after-statement -Wc++-compat -Wwrite-strings util.c
util.c: In function 'Perl_safesysrealloc':
util.c:277:12: warning: pointer 'where' may be used after 'realloc' [-Wuse-after-free]
  277 |         UV was_where = PTR2UV(where);
      |            ^~~~~~~~~
util.c:291:15: note: call to 'realloc' here
  291 |         ptr = (Malloc_t)PerlMem_realloc(where,size);
```
I'm still getting a `-Wuse-after-free` warning, albeit at a different line in `util.c`.

My previous report was compiled with `gcc12`.  Somehow that version got wiped out when I went to add `gcc13`, so I acknowledge that I'm not testing under precisely the same conditions.  However, I took the precaution of building both blead and the branch on Linux with `gcc-12` as well.  The branch still emits both warnings.

**Linux**

*blead*
```
$ report-build-warnings db91d049e0.linux.threaded.debugging.gcc-12.makeminiperl.output.txt.gz 
File:  db91d049e0.linux.threaded.debugging.gcc-12.makeminiperl.output.txt.gz

  Warray-bounds                              1
  Wuse-after-free                            1

$ parse-one-warnings-group db91d049e0.linux.threaded.debugging.gcc-12.makeminiperl.output.txt.gz Wuse-after-free
File:  db91d049e0.linux.threaded.debugging.gcc-12.makeminiperl.output.txt.gz

[
  {
    char   => 17,
    group  => "Wuse-after-free",
    line   => 329,
    source => "util.c",
    text   => "pointer \xE2\x80\x98where\xE2\x80\x99 may be used after \xE2\x80\x98realloc\xE2\x80\x99",
  },
]
```
*branch*
```
$ report-build-warnings db91d049e0.linux.threaded.debugging.gcc-12.makeminiperl.leonerd.output.txt.gz 
File:  db91d049e0.linux.threaded.debugging.gcc-12.makeminiperl.leonerd.output.txt.gz

  Warray-bounds                              1
  Wuse-after-free                            1

$ parse-one-warnings-group db91d049e0.linux.threaded.debugging.gcc-12.makeminiperl.leonerd.output.txt.gz Wuse-after-free
File:  db91d049e0.linux.threaded.debugging.gcc-12.makeminiperl.leonerd.output.txt.gz

[
  {
    char   => 17,
    group  => "Wuse-after-free",
    line   => 329,
    source => "util.c",
    text   => "pointer \xE2\x80\x98where\xE2\x80\x99 may be used after \xE2\x80\x98realloc\xE2\x80\x99",
  },
]
```

--- a/cflags.SH
+++ b/cflags.SH
@@ -379,6 +379,19 @@ esac
 ;;
 esac
 
+# gcc version 12 and 13 are overly aggressive with use-after-free warnings
+# and have false positives on code that shouldn't warn, and they haven't
+# sorted it out yet. See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=108115
+case "$gccversion" in
+"1"[23]*)
+    for f in -Wno-use-after-free
+    do
+        echo "cflags.SH: Adding $f because of false positives in gccversion '$gccversion'"
+        warn="$warn $f"
+    done
+;;
+esac
+
 echo "cflags.SH: cc       = $cc"
 echo "cflags.SH: ccflags  = $ccflags"
 echo "cflags.SH: stdflags = $stdflags"

